storage/src/main/java/org/apache/kafka/storage/internals/log/LogSegment.java

```diff
@@ -0,0 +1,873 @@
+                LOGGER.info("Failed to delete {} {} because it does not exist.", fileType, file.getAbsolutePath());
+            return null;
+        } catch (IOException e) {
+            throw new IOException("Delete of " + fileType + " " + file.getAbsolutePath() + " failed.", e);
+        }
+    }
+
+    public boolean deleted() {
+        return !log.file().exists() && !offsetIndexFile().exists() && !timeIndexFile().exists() && !txnIndex.file().exists();
+    }
+
+
+    /**
+     * The last modified time of this log segment as a unix time stamp
+     */
+    public long lastModified() {
+        return log.file().lastModified();
+    }
+
+    /**
+     * The largest timestamp this segment contains, if maxTimestampSoFar >= 0, otherwise None.
+     */
+    public OptionalLong largestRecordTimestamp() {
+        long maxTimestampSoFar = maxTimestampSoFar();
+        if (maxTimestampSoFar >= 0)
+            return OptionalLong.of(maxTimestampSoFar);
+        return OptionalLong.empty();
+    }
+
+    /**
+     * The largest timestamp this segment contains.
+     */
+    public long largestTimestamp() {
+        long maxTimestampSoFar = maxTimestampSoFar();
+        if (maxTimestampSoFar >= 0)
+            return maxTimestampSoFar;
+        return lastModified();
+    }
+
+    /**
+     * Change the last modified time for this log segment
+     */
+    public void setLastModified(long ms) throws IOException {
+        FileTime fileTime = FileTime.fromMillis(ms);
+        Files.setLastModifiedTime(log.file().toPath(), fileTime);
+        Files.setLastModifiedTime(offsetIndexFile().toPath(), fileTime);
+        Files.setLastModifiedTime(timeIndexFile().toPath(), fileTime);
+    }
+
+    public static LogSegment open(File dir, long baseOffset, LogConfig config, Time time, int initFileSize, boolean preallocate) throws IOException {
+        return open(dir, baseOffset, config, time, false, initFileSize, preallocate, "");
+    }
+
+    public static LogSegment open(File dir, long baseOffset, LogConfig config, Time time, boolean fileAlreadyExists,
+                                  int initFileSize, boolean preallocate, String fileSuffix) throws IOException {
+        int maxIndexSize = config.maxIndexSize;
+        return new LogSegment(
+            FileRecords.open(LogFileUtils.logFile(dir, baseOffset, fileSuffix), fileAlreadyExists, initFileSize, preallocate),
+            LazyIndex.forOffset(LogFileUtils.offsetIndexFile(dir, baseOffset, fileSuffix), baseOffset, maxIndexSize),
+            LazyIndex.forTime(LogFileUtils.timeIndexFile(dir, baseOffset, fileSuffix), baseOffset, maxIndexSize),
+            new TransactionIndex(baseOffset, LogFileUtils.transactionIndexFile(dir, baseOffset, fileSuffix)),
+            baseOffset,
+            config.indexInterval,
+            config.randomSegmentJitter(),
+            time);
+    }
+
+    public static void deleteIfExists(File dir, long baseOffset, String fileSuffix) throws IOException {
```

Comment:
I noticed that you didn't implement the case where fileSuffix is not provided because we are always providing a suffix. Nice!